src/endf.cpp

```diff
@@ -0,0 +1,261 @@
+
+namespace openmc {
+
+//==============================================================================
+// Functions
+//==============================================================================
+
+Interpolation int2interp(int i)
+{
+  // TODO: We are ignoring specification of two-dimensional interpolation
+  // schemes (method of corresponding points and unit base interpolation). Those
+  // should be accounted for in the distribution classes somehow.
+
+  switch (i) {
+  case 1: case 11: case 21:
+    return Interpolation::histogram;
+  case 2: case 12: case 22:
+    return Interpolation::lin_lin;
+  case 3: case 13: case 23:
+    return Interpolation::lin_log;
+  case 4: case 14: case 24:
+    return Interpolation::log_lin;
+  case 5: case 15: case 25:
+    return Interpolation::log_log;
+  default:
+    throw std::runtime_error{"Invalid interpolation code."};
+  }
+}
+
+bool is_fission(int mt)
+{
+  return mt == 18 || mt == 19 || mt == 20 || mt == 21 || mt == 38;
+}
+
+bool is_disappearance(int mt)
+{
+  if (mt >= N_DISAPPEAR && mt <= N_DA) {
+    return true;
+  } else if (mt >= N_P0 && mt <= N_AC) {
+    return true;
+  } else if (mt == N_TA || mt == N_DT || mt == N_P3HE || mt == N_D3HE
+    || mt == N_3HEA || mt == N_3P) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+bool is_inelastic_scatter(int mt)
+{
+  if (mt < 100) {
+    if (is_fission(mt)) {
+      return false;
+    } else {
+      return mt >= MISC && mt != 27;
+    }
+  } else if (mt <= 200) {
+    return !is_disappearance(mt);
+  } else if (mt >= N_2N0 && mt <= N_2NC) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+std::unique_ptr<Function1D>
+read_function(hid_t group, const char* name)
+{
+  hid_t dset = open_dataset(group, name);
+  std::string func_type;
+  read_attribute(dset, "type", func_type);
+  std::unique_ptr<Function1D> func;
+  if (func_type == "Tabulated1D") {
+    func = std::make_unique<Tabulated1D>(dset);
+  } else if (func_type == "Polynomial") {
+    func = std::make_unique<Polynomial>(dset);
+  } else if (func_type == "CoherentElastic") {
+    func = std::make_unique<CoherentElasticXS>(dset);
+  } else if (func_type == "IncoherentElastic") {
+    func = std::make_unique<IncoherentElasticXS>(dset);
+  } else {
+    throw std::runtime_error{"Unknown function type " + func_type +
+      " for dataset " + object_name(dset)};
+  }
+  close_dataset(dset);
+  return func;
+}
+
+//==============================================================================
+// Polynomial implementation
+//==============================================================================
+
+Polynomial::Polynomial(hid_t dset)
+{
+  // Read coefficients into a vector
+  read_dataset(dset, coef_);
+}
+
+double Polynomial::operator()(double x) const
+{
+  // Use Horner's rule to evaluate polynomial. Note that coefficients are
+  // ordered in increasing powers of x.
+  double y = 0.0;
+  for (auto c = coef_.crbegin(); c != coef_.crend(); ++c) {
```

Comment:
Optional: If possible, would recommend making `c` const, similar to the for loop on line `140`.